### ISSUE TITLE
Plan for 2.0.0 

### ISSUE THREAD

**Issue by chpxu** [chpxu/vscode-firefox-quantum-themes#45]
# Motivation
I have since changed my ideas significantly from #23 when I first brought up a "Component-Based File System". The main principles and motivation behind this remain the same
- Make it easier to maintain:
  - No more manually changing colours. Colours are stored as variables and the only thing that needs to change is the attribute you want to theme
  - Automatically generated. Change one bit of JavaScript and new theme files are made
This is also an excuse for me to overly-engineer a theme extension. It also may be useful to generate other themes from scratch by simply changing colours.

# Changes
I will change the structure of the repo slightly with the addition of more files, namely the new structure will be similar to:
```
[Root Dir]
|> screenshots
|> themes
|> scripts
  | -- colours.ts
  | -- generate.ts
  | -- generator.ts
|> <other files>
```
To explain the purpose of the TypeScript files:
1. `generator.ts` contains the interfaces and functions that generate a block of object code and append it to global `ql` and `qd` ('quantum light' and 'quantum dark' respectively) objects.
2. `colours.ts` is simply a file containing an object with all the colours, both light and dark themes as recorded in Mozilla's (Firefox DevTools colours)[https://developer.mozilla.org/en-US/docs/Tools/DevToolsColors]. This file also contains the arrays of attributes to be themed. As such, most editing will take place here.
3. `generate.ts` is the main file of execution. It  runs the appropriate functions in `generator.ts`, appends them to the theme objects and then creates and writes the JSON files for the themes. 

This change results in the following consequence
1. The `generate.ts` file is the  only file of execution. One funcgtion logic is finalised, it shouldn't need to be changed much. 
3. `generator.ts` will also be rarely updated, once function logic is finalised.
4. `colours.ts` will be the main file of editing as it contains the arrays of attributes to be themed.

## Possible Issues
1. With large theme files, wouldn't it take long to generate the files? The effect would only come noticeably when you have thousands of entries, is my answer. Furthermore, the user never generates theme files themselves. They come pre-generated and act as like regular `theme.json` files. Running the script is only for development purposes. Currently though, everything elaready takes about 2.4-2.7s to create on my machine.
2. Lack of organisation in the `json` theme files. I ideally want to sort the attributes alphabetically (which puts them into the correct categories thanks to the VSCode's team naming system). This shouldn't be too hard to implement but isn't the focus this early on in development, but it should be accomplished by 1 function.
  - Currently, the JSON file is just appropriately indented.

This will be the main development going forward. Version 1.2.0 is expected to be the last minor version of the 1.x branch. You can see 2.0.0 development progress by going to branches/2.0.0

# Checklist
In the past I haven't been very open about what I'm doing. This time, I want to communicate where I am for this update
## Alpha Stages
The alpha stages is simply the main implementation existing
- [x] Make `colours.ts` containing variables which are arrays, each with the respective colour from Dark and Light
  - [x]  All attributes to be themed
- [x] Create and fill in `generate.ts`
  - [x] import appropriate files
  - [x] Use `fs` to create theme files
  - [x] Functions to generate theme files complete
   
- [x] Create `generator.ts`, the file contains the logic to generate an object block 
  - [x] A function `generateColours` to generate a `key: value` pair for an attribute
  - [x] A function `generateTokenColours` to generate a block of object code
  - [x] Both functions can append to a global object.
## Beta
With this stage, the feature set should be on par with stable. This now allows time to actually develop the theme further and improve the repo.
- [x] Ensure feature parity with current stable (currently `v1.2.0`).
- [x] Begin going through default supported languages and checking the theme is correct.
   - [x] Check language is themed as best as possible witu default scopes.
   - [x] If an extension provides a LSP for a language, check it is also correctly themed.
- [x] Repeat the same as above but for languages of my own choosing.
- [x] Attempt to organise the theme files better.
  - [x] Sorting everything by category and alphabetically in `colors`.
  - [x] Sorting everything in the order of global, then language-specific.
- [x] Look at VSCode docs to see what new themes can be added for the UI.

## RC
- [ ] Update SCREENSHOTS for languages and the theme. These are severely out of date.
- [ ] Final checks to ensure theme isn't broken in any way.

## Release as 2.0.0
